src/libmongoc/src/mongoc/mongoc-topology-description.c

```diff
@@ -895,8 +989,12 @@ mongoc_topology_description_select (
    _mongoc_array_init (&suitable_servers,
```

Comment:
Tasks on pre 5.0 sharded servers are failing: https://spruce.mongodb.com/version/61e76013a4cf4705407cfa1f/tasks?page=0&sorts=STATUS%3AASC

This reproduces locally with this environment:
```
mlaunch init --replicaset --nodes=1 --sharded=1 --mongos=2 --priority --binarypath $(m bin 4.2-ent) --hostname localhost --enableMajorityReadConcern --setParameter enableTestCommands=1 --dir .menv_sh2_42
```

I was unable to fully track down the cause, but I have an idea.

libmongoc sets the initial topology type to `MONGOC_TOPOLOGY_SINGLE` when given a URI with one host and no replicaSet option (e.g. `mongodb://foo.example.com/`) . That behavior is driver dependent and noted in [this spreadsheet](https://docs.google.com/spreadsheets/d/1sddTIxDtaB_x6z8c2vp6QbTwJSHm5nEPCRHbhUBLbPs/edit#gid=0).

I applied this diff:
```patch
diff --git a/src/libmongoc/src/mongoc/mongoc-topology-description.c b/src/libmongoc/src/mongoc/mongoc-topology-description.c
index 2ba90466c..0f9a5fcb7 100644
--- a/src/libmongoc/src/mongoc/mongoc-topology-description.c
+++ b/src/libmongoc/src/mongoc/mongoc-topology-description.c
@@ -978,6 +978,11 @@ mongoc_topology_description_select (
    if (topology->type == MONGOC_TOPOLOGY_SINGLE) {
       sd = mongoc_set_get_item_const (mc_tpld_servers_const (topology), 0);
 
+      if (optype == MONGOC_SS_AGGREGATE_WITH_WRITE &&
+          sd->max_wire_version < WIRE_VERSION_5_0) {
+         *must_use_primary = true;
+      }
+
       if (sd->has_hello_response) {
          RETURN (sd);
       } else {
```

Which resolves the failure locally.

Reply:
This is a good catch. I had wrongly assumed those failures were spurious. I'll add those changes.